cpp/src/arrow/c/abi.h

```diff
@@ -106,6 +212,98 @@ struct ArrowArrayStream {
 
 #endif  // ARROW_C_STREAM_INTERFACE
 
+#ifndef ARROW_C_DEVICE_STREAM_INTERFACE
+#define ARROW_C_DEVICE_STREAM_INTERFACE
+
+/// \brief Equivalent to ArrowArrayStream, but for ArrowDeviceArrays.
+///
+/// This stream is intended to provide a stream of data on a single
+/// device, if a producer wants data to be produced on multiple devices
+/// then multiple streams should be provided. One per device.
+struct ArrowDeviceArrayStream {
+  /// \brief The device that this stream produces data on.
+  ///
+  /// All ArrowDeviceArrays that are produced by this
+  /// stream should have the same device_type as set
+  /// here. The device_type needs to be provided here
+  /// so that consumers can provide the correct type
+  /// of queue_ptr when calling get_next.
+  ArrowDeviceType device_type;
+
+  /// \brief Callback to get the stream schema
+  /// (will be the same for all arrays in the stream).
+  ///
+  /// If successful, the ArrowSchema must be released independantly from the stream.
+  /// The schema should be accessible via CPU memory.
+  ///
+  /// \param[in] self The ArrowDeviceArrayStream object itself
+  /// \param[out] out C struct to export the schema to
+  /// \return 0 if successful, an `errno`-compatible error code otherwise.
+  int (*get_schema)(struct ArrowDeviceArrayStream* self, struct ArrowSchema* out);
+
+  /// \brief Callback to get the device id for the next array.
+  ///
+  /// This is necessary so that the proper/correct stream pointer can be provided
+  /// to get_next.
+  ///
+  /// The next call to `get_next` should provide an ArrowDeviceArray whose
+  /// device_id matches what is provided here, and whose device_type is the
+  /// same as the device_type member of this stream.
```

Comment:
I'm not certain I follow.  Isn't the `ArrowDeviceArray` passed to `get_next` an "out" parameter?  Are you saying that the `ArrowDeviceArray` struct itself (not the buffers) needs to be allocated on the device?

Reply:
no, I was referring to the `device_id` member and `device_type` member that get populated in the `ArrowDeviceArray` that is returned from `get_next`

Reply:
This is a weird API choice, all because you want the consumer to pass its CUDA stream of choice...

Reply:
ultimately this is a consequence of the fact that the existing frameworks and APIs don't provide any good way to manage the stream's lifetime easily which makes having the consumer pass the stream be the safest route to take.

I'm absolutely open to suggestions to make this better as long as the consumer is able to pass in the desired stream.

Reply:
> the existing frameworks and APIs don't provide any good way to manage the stream's lifetime easily

What do you mean by that? Would you care to give a more concrete example? For example CUDA allows you to destroy a stream:
https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1gfda584f1788ca983cb21c5f4d2033a62

Reply:
Many frameworks manage their streams internally and don't expose those streams in their APIs, so there isn't a clean way to share or release ownership of a stream in a way that a stream can be guaranteed to not be destroyed until the release callback is called.

See the conversations here:
- https://github.com/dmlc/dlpack/issues/57
- https://github.com/data-apis/consortium-feedback/issues/1
- https://github.com/data-apis/array-api/pull/106

Reply:
>     * [Specify synchronization semantics dmlc/dlpack#57](https://github.com/dmlc/dlpack/issues/57)

I tried to read through this and did not find a convincing argument that we shouldn't let the producer expose the original stream value. The entire discussion seemed to revolve around API convenience, not correctness. Also, people there seem to agree that the CUDA Array Interface from Numba is sane.

Did I miss something?

Reply:
There's a lot of discussion in that issue related to internal stream handling under contexts and schedulers and similar terms. It all boils down to the same discussion of many producers being unable to release or share ownership of their streams.

I think this comment does a good job of summarizing the options that were considered: https://github.com/dmlc/dlpack/issues/57#issuecomment-753220425

And then this comment summarizes discussion of those options: https://github.com/dmlc/dlpack/issues/57#issuecomment-774482973


The lifetime management of streams as defined by the Numba documentation for `__cuda_array_interface__` (https://numba.readthedocs.io/en/stable/cuda/cuda_array_interface.html#streams) requires that keeping the object (typically array) that produces the `__cuda_array_interface__` alive also keeps the stream alive. In most cases libraries don't associate a stream with the object since it's valid to use multiple streams with a single object.

Here's the current state of things across a handful of projects:
- Numba
  - Handles stream lifetime properly for `__cuda_array_interface__` since they store a stream object as part of their array: https://github.com/numba/numba/blob/008077553b558bd183668ecd581d4d0bc54bd32c/numba/cuda/cudadrv/devicearray.py#L119-L139
  - Doesn't support `__dlpack__`
- CuPy
  - Doesn't implement stream lifetime properly for `__cuda_array_interface__` as far as I can tell, where they just get a stream ptr as an integer and if it's not the default stream someone could change the current stream ptr and end up having it be destroyed out from underneath it: https://github.com/cupy/cupy/blob/c92d5bc16293300297b843b4ebb364125697c131/cupy/_core/core.pyx#L258-L262 (cc @leofang)
  - Handles `__dlpack__` properly: https://github.com/cupy/cupy/blob/c92d5bc16293300297b843b4ebb364125697c131/cupy/_core/core.pyx#L285-L327
- PyTorch
  - Doesn't support passing the stream in `__cuda_array_interface__` currently: https://github.com/pytorch/pytorch/blob/def50d253401540cfdc6c0fffa444d0ee643cc11/torch/_tensor.py#L1005-L1066
  - Handles `__dlpack__` properly: https://github.com/pytorch/pytorch/blob/def50d253401540cfdc6c0fffa444d0ee643cc11/torch/_tensor.py#L1345
- Tensorflow and JAX
  - Doesn't support passing the stream in `__cuda_array_interface__` currently: https://github.com/tensorflow/tensorflow/blob/ea6a0f282d2b7ce20891dfc24ec8fe107eeaf22d/tensorflow/compiler/xla/python/py_buffer.cc#L254-L300
  - Doesn't support `__dlpack__`, but has explicit `to` and `from` functions for using dlpack pycapsules but does not handle streams: https://github.com/tensorflow/tensorflow/blob/6a050b6c15ed2a545693bc171f5e95dacbe05839/tensorflow/compiler/xla/python/dlpack.cc#L285-L363
- cuDF
  - Doesn't support passing the stream in `__cuda_array_interface__` currently: https://github.com/rapidsai/cudf/blob/50718e673ff53b18706cf66c6e02cda8e30681fe/python/cudf/cudf/core/column/numerical.py#L169-L191
  - Doesn't support `__dlpack__` but has explicit `to` and `from` functions for using dlpack pycapsules but does not handle streams: https://github.com/rapidsai/cudf/blob/50718e673ff53b18706cf66c6e02cda8e30681fe/cpp/src/interop/dlpack.cpp#L218-L294

Reply:
cc @rgommers @leofang @tqchen (please feel free to include anyone else) do you happen to know if there was any other discussion captured that could be linked here regarding the decision to have a consumer hand a stream to the producer?

Reply:
Thanks the pointers.

> I think this comment does a good job of summarizing the options that were considered: [dmlc/dlpack#57 (comment)](https://github.com/dmlc/dlpack/issues/57#issuecomment-753220425)

Yes, I read this. It looks like solution S1, which is also the one I'm proposing, is considered the most flexible (I don't understand the "harder for compilers" comment, though).

> And then this comment summarizes discussion of those options: [dmlc/dlpack#57 (comment)](https://github.com/dmlc/dlpack/issues/57#issuecomment-774482973)

I read this too, but it doesn't actually mention S1, for reasons I wasn't able to understand.

>  In most cases libraries don't associate a stream with the object since it's valid to use multiple streams with a single object.

But you have to actually synchronize on the right stream before being able to use the object, right? How does the user know which stream to synchronize on, if they didn't produce the data themselves?

Reply:
> Yes, I read this. It looks like solution S1, which is also the one I'm proposing, is considered the most flexible (I don't understand the "harder for compilers" comment, though).

From: https://github.com/dmlc/dlpack/issues/57#issuecomment-751715168
> It also brings extra burden to the compilers themselves. The compiler will need to generate optional synchronization code based on the streams, which is non-trivial.

> But you have to actually synchronize on the right stream before being able to use the object, right?

Something / someone needs to guarantee that there isn't a data race with regards to using multiple non-blocking streams, yes. That could be done with events, stream synchronization, or device synchronization.

> How does the user know which stream to synchronize on, if they didn't produce the data themselves?

If you're staying within your framework / library then the expectation is for the framework / library to handle things for the user. If crossing framework / library boundaries, then the expectation is to be reliant on things like interchange protocols to handle the synchronization semantics.

Reply:
> do you happen to know if there was any other discussion captured that could be linked here regarding the decision to have a consumer hand a stream to the producer

Sorry I wasn't able to respond promptly. Is the question still open?

In the case of CAI, [it is required](https://numba.readthedocs.io/en/stable/cuda/cuda_array_interface.html#streams) that someone handles the exporting stream's lifetime properly:
> Like data, CUDA streams also have a finite lifetime. It is therefore required that a Producer exporting data on the interface with an associated stream ensures that the exported stream’s lifetime is equal to or surpasses the lifetime of the object from which the interface was exported.

and this was considered a burden when discussing the DLPack support. A few libraries like Numba, for example, had to [hold the reference to the underlying stream](https://numba.readthedocs.io/en/stable/cuda/cuda_array_interface.html#producing-arrays). I believe this was the main concern for DLPack to place the requirement on the consumer instead of the producer.